src/qml/qml_owned_ptr.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#ifdef MIXXX_DEBUG_ASSERTIONS_ENABLED
+#include <QObject>
+#include <QQmlEngine>
+#endif
+#include <QPointer>
+#include <memory>
+
+#include "util/assert.h"
+
+// Use this wrapper class to clearly represent a raw pointer that is owned by a
+// QML Engine. Objects which derive from QObject, have their lifetime governed
+// by the QML (or JavaScript) Engine, and thus such pointers do not require a
+// manual delete to free the heap memory when they go out of scope, as they will
+// be handled by the engine garbage collector.
+template<typename T>
+class qml_owned_ptr final {
+  public:
+    qml_owned_ptr() noexcept
+            : m_ptr{nullptr} {
+    }
+
+    qml_owned_ptr(std::nullptr_t) noexcept
+            : m_ptr{nullptr} {
+    }
+
+    explicit qml_owned_ptr(T* t) noexcept
+            : m_ptr{t} {
+        DEBUG_ASSERT(m_ptr);
+        QQmlEngine::setObjectOwnership(m_ptr, QQmlEngine::JavaScriptOwnership);
+    }
+
+    // explicitly generate trivial destructor (since decltype(m_ptr) is not a class type)
+    ~qml_owned_ptr() noexcept = default;
+
+    // Rule of 5
+    qml_owned_ptr(const qml_owned_ptr<T>&) = delete;
+    qml_owned_ptr& operator=(const qml_owned_ptr<T>&) = delete;
+    qml_owned_ptr(const qml_owned_ptr<T>&& other) = delete;
+    qml_owned_ptr& operator=(const qml_owned_ptr<T>&& other) = delete;
+
+    // If U* is convertible to T* then qml_owned_ptr<U> is convertible to qml_owned_ptr<T>
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    qml_owned_ptr(qml_owned_ptr<U>&& u) noexcept
+            : m_ptr{u.m_ptr} {
+        u.m_ptr = nullptr;
+    }
+
+    // If U* is convertible to T* then qml_owned_ptr<U> is assignable to qml_owned_ptr<T>
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    qml_owned_ptr& operator=(qml_owned_ptr<U>&& u) noexcept {
+        qml_owned_ptr temp{std::move(u)};
+        std::swap(temp.m_ptr, m_ptr);
```

Comment:
not sure about the `temp` here. I loosely remember it being required to be exception safe (IIRC something like this was recommended on cppreference), but I'm not sure. Can you double check and maybe leave a comment?

Reply:
Yeah, okay. So after a bit of googling and AI, I found the source: https://www.modernescpp.com/index.php/the-copy-and-swap-idiom

I guess its not required in this case, but better keep it anyways.